### PR TITLE
Introduce typed version of coalesce

### LINKED ISSUE #1376: Performing coalesce on NumberPath in from or where clause
`ComparableExpressionBase`'s `public final Coalesce<T> coalesce(Expression<?>...exprs)` returns a `Coalesce<T>`. `Coalesce` does not have `SimpleExpression`'s `public BooleanExpression eq(T right)` method, so I assume the idea is to use `Coalesce`'s `public NumberExpression<?> asNumber()`. However if you try to E.G: (I know the query is nonsense)

``` java
query()
.from(someRelationalPathBase)
.innerJoin(someOtherRelationalPathBase)
.on(someRelationalPathBase.id.coalesce(someOtherRelationalPathBase.id).asNumber().eq(someRelationalPathBase.id));
```

you receive the compiler error `The method eq(capture#2-of ?) in the type SimpleExpression<capture#2-of ?> is not applicable for the arguments (NumberPath<Long>)`. I'm working around this with the following code:

``` java
 @SuppressWarnings("unchecked")
 public static <T extends Number & Comparable<T>> NumberOperation<T> coalesceNumbers(NumberPath<T> arg0, NumberPath<T> arg1) {
   return ((NumberOperation<T>) arg0.coalesce(arg1).asNumber());
 }
```

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1376

### CODE REVIEW

**Comment by timowest** [comment]
This is more of a sketch at the moment. The signature clashes with the old version of `coalesce`.

The new version provides better return types, but different name should be used to avoid clashes.

**Comment by robertandrewbain** [comment]
Looks good to me, even as a sketch.

File: querydsl-core/src/main/java/com/querydsl/core/types/dsl/BooleanOperation.java [hunk 1]
```diff
@@ -68,4 +68,28 @@ public BooleanExpression not() {
...
+
+    @Override
+    public BooleanExpression coalesce(SimpleExpression<Boolean> other) {
+        return Expressions.booleanOperation(Ops.COALESCE, Expressions.list(this, other));
+    }
+
+    @Override
+    public BooleanExpression coalesce(SimpleExpression<Boolean>... exprs) {
+        if (exprs.length == 0) {
+            return this;
```
Comment (robertandrewbain) [inline querydsl-core/src/main/java/com/querydsl/core/types/dsl/BooleanOperation.java]: Throw `IllegalArgumentException`?

**Comment by robertandrewbain** [comment]
I'd expect to see some unit tests for this.

**Pull request closed without merging**